rfcs/20180731-dockerfile-assembler.md

```diff
@@ -0,0 +1,164 @@
+Dockerfiles that get committed into source control. These Dockerfiles are fully
+documented and support argument customization. Unlike the parameterized image
+builder script, this system excludes the image deployment steps, which should
+be handled by a totally different system anyway.
+
+This section lightly describes the design, which is fully implemented in [this
+pull request to the main TensorFlow
+repo](https://github.com/tensorflow/tensorflow/pull/21291).
+
+Partial files are syntactically valid but incomplete files with Dockerfile
+syntax.
+
+Assembly is controlled by a specification file, defined in yaml. The spec
+defines the partials, the ARGs they use, the list of Dockerfiles to generate
+based on ordered lists of partials, and documentation for those values.
+
+The assembler is a python script that accepts a spec and generates a bunch of
+Dockerfiles to be committed. The spec includes documentation and descriptions,
+and the output Dockerfiles are fully documented and can be built manually.
+
+# Impact
+
+This approach has many convenient benefits:
+
+*   The result is concrete, buildable, documented Dockerfiles. Users who wish
+to build their own images locally do not need to also understand the build
+system.
+*   This implementation is agnostic to what images we would like to make
+available online (i.e. our Docker story). It's very easy to add new dockerfile
+outputs.
+*   The build-test-and-deploy-images process is decoupled from the Dockerfile
+generation process.
+*   Control of the set of dockerfiles is centralized to the spec file, instead
+of being spread across each Dockerfile.
+*   The spec can be extended to add more conveniences. My implementation, for
+example, already includes de-duplication of many similar Dockerfile
+specifications.
+*   All dockerfiles are consistently documented.
+*   Common pieces of code, like a slick shell environment or a Jupyter
+interface, can be updated in batch by updating a single partial file.
+*   The spec can also be used in the image building process, e.g. to read all
+available args.
+
+# Caveats and Rejected Alternatives
+
+I considered two alternatives while working on this.
+
+## Modern Multi-Stage Dockerfile
+
+"Multi-stage Building" is a powerful new Dockerfile feature that supports
+multiple FROM statements in one Dockerfile. It is meant to be used for creating
+artifacts with one image before using those artifacts in another image, but you
+can also use variable expansion in any FROM line:
+
+
+```dockerfile
+# If --build-arg FROM_FOO is set, build from foo. else build from bar.
+ARG FROM_FOO
+ARG _HELPER=${FROM_FOO:+foo}
+ARG BASE_IMAGE=${_HELPER:-bar}
+FROM ${BASE_IMAGE}
+…
+```
+
+...which means that you can dynamically set multiple FROM images. My first
+draft used ARGs and FROMs in a single Dockerfile to manipulate build stages.
+[The resulting
+Dockerfile](https://gist.github.com/angersson/3d2b5ae6a01de4064b1c3fe7a56e3821)
```

Comment:
The Dockerfile in this gist is truncated.
Also, I'm not sure why ARGs and multi-stage builds need to work together. 

ARGs is for templating, allowing you to have a generic Dockerfile.
Multi-stage build is to avoid shipping with your build dependencies.

So I'm confused, for me there are totally different concepts.

Actually, when I took a quick look at the TF Dockerfiles, I thought multi-stage builds were not a good fit because:
- devel images need to maintain those build dependencies
- non-devel images already install from pip packages and don't have build dependencies (python deps excluded)
- many dependencies are installed through the package manager (`apt-get`) which doesn't work with multi-stage builds anyway

Reply:
I think I may have been in the middle of updating the file when I decided to drop that approach. I fixed the line.

I've revised this entire section since your comment (which illustrated that this portion of the doc was poorly explained, thanks) to clarify why I included this explanation. Regular multi-stage builds wouldn't be very useful at all, but the support for multiple FROMs that they added would let us do very evil things that don't work very well in the long run. I included the gist to show how bad such a Dockerfile can get (it's not meant to be useful aside from that).